File: scripts/test_farnell_regions.py
#!/usr/bin/env python3
"""
Farnell API Region & Search Method Test

Tests Farnell API with different regions and search methods to understand empty results.
"""

import os
import sys
import json
import requests
from urllib.parse import urlencode

# Load credentials from environment or .env file
try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass

FARNELL_API_KEY = os.getenv('FARNELL_API_KEY')

if not FARNELL_API_KEY:
    print("❌ Error: FARNELL_API_KEY must be set in environment")
    print("   Add it to .env file or export it:")
    print("   export FARNELL_API_KEY='your_api_key'")
    sys.exit(1)

print(f"🔑 Farnell API Key: {FARNELL_API_KEY[:10]}...{FARNELL_API_KEY[-4:]}")
print()


# Different Farnell/Element14 regions
REGIONS = [
    {
        'name': 'UK Farnell',
        'base': 'https://uk.farnell.com',
        'storeId': 10164,
        'catalogId': 15001,
        'langId': 44
    },
    {
        'name': 'US Newark',
        'base': 'https://www.newark.com',
        'storeId': 10194,
        'catalogId': 15001,
        'langId': -1
    },
    {
        'name': 'Element14 (Global)',
        'base': 'https://www.element14.com',
        'storeId': 10159,
        'catalogId': 15001,
        'langId': -1
    },
    {
        'name': 'CPC UK',
        'base': 'https://cpc.farnell.com',
        'storeId': 10179,
        'catalogId': 15001,
        'langId': 44
    }
]


def test_webapp_search(region, mpn, search_params):
    """Test webapp/wcs/stores/servlet/Search endpoint"""
    
    url = f"{region['base']}/webapp/wcs/stores/servlet/Search"
    
    params = {
        'storeId': region['storeId'],
        'catalogId': region['catalogId'],
        'langId': region['langId'],
        'apiKey': FARNELL_API_KEY,
        'resultsPerPage': 5,
        **search_params
    }
    
    try:
        response = requests.get(url, params=params, timeout=15)
        
        result = {
            'status': response.status_code,
            'url': response.url,
            'success': False,
            'num_results': 0,
            'products': []
        }
        
        if response.status_code == 200:
            try:
                data = response.json()
                
                # Try different response structures
                if 'premierFarnellPartNumberReturn' in data:
                    pfpn = data['premierFarnellPartNumberReturn']
                    result['num_results'] = pfpn.get('numberOfResults', 0)
                    result['products'] = pfpn.get('products', [])
                elif 'products' in data:
                    result['num_results'] = len(data['products'])
                    result['products'] = data['products']
                
                result['success'] = result['num_results'] > 0
                result['raw'] = data
                
            except Exception as e:
                result['error'] = f"JSON parse error: {e}"
                result['raw_text'] = response.text[:500]
        else:
            result['error'] = response.text[:500]
        
        return result
        
    except Exception as e:
        return {
            'status': 0,
            'success': False,
            'error': str(e)
        }


def test_rest_api(mpn, search_method):
    """Test REST API v3 endpoint"""
    
    base_url = 'https://api.element14.com/catalog/products'
    
    params = {
        'callInfo.apiKey': FARNELL_API_KEY,
        'storeInfo.id': 'uk.farnell.com',
        'resultsSettings.numberOfResults': 5,
        'resultsSettings.responseGroup': 'medium'
    }
    
    if search_method == 'keyword':
        params['term'] = mpn
    elif search_method == 'mpn_exact':
        params['term'] = f'manuPartNum:{mpn}'
    elif search_method == 'mpn_field':
        params['manufacturerPartNumber'] = mpn
    
    try:
        response = requests.get(base_url, params=params, timeout=15)
        
        result = {
            'status': response.status_code,
            'url': response.url,
            'success': False,
            'num_results': 0,
            'products': []
        }
        
        if response.status_code == 200:
            try:
                data = response.json()
                
                # Check different response structures
                if 'products' in data:
                    result['products'] = data['products']
                    result['num_results'] = len(data['products'])
                elif 'productSearchReturn' in data:
                    products = data['productSearchReturn'].get('products', [])
                    result['products'] = products
                    result['num_results'] = len(products)
                
                result['success'] = result['num_results'] > 0
                result['raw'] = data
                
            except Exception as e:
                result['error'] = f"JSON parse error: {e}"
                result['raw_text'] = response.text[:500]
        else:
            result['error'] = response.text[:500]
        
        return result
        
    except Exception as e:
        return {
            'status': 0,
            'success': False,
            'error': str(e)
        }


def print_result(title, result):
    """Pretty print test result"""
    
    print(f"\n   {title}")
    print(f"   Status: {result['status']}")
    
    if result['success']:
        print(f"   ✅ SUCCESS - Found {result['num_results']} product(s)")
        
        if result['products']:
            product = result['products'][0]
            print(f"\n   First product:")
            
            # Try different field names
            mpn = (product.get('translatedManufacturerPartNumber') or 
                   product.get('manufacturerPartNumber') or 
                   product.get('mpn') or 
                   'N/A')
            sku = product.get('sku') or product.get('id') or 'N/A'
            desc = (product.get('displayName') or 
                    product.get('productDescription') or 
                    product.get('description') or 
                    'N/A')
            
            print(f"      MPN: {mpn}")
            print(f"      SKU: {sku}")
            print(f"      Description: {desc[:50]}...")
    else:
        if 'error' in result:
            print(f"   ❌ FAILED - {result['error']}")
        else:
            print(f"   ⚠️ No results found")


def test_region(region, mpn):
    """Test all search methods for a specific region"""
    
    print(f"\n{'='*80}")
    print(f"📍 {region['name']} - {mpn}")
    print(f"{'='*80}")
    
    search_methods = [
        {
            'name': 'Keyword Search (st)',
            'params': {'st': mpn, 'searchType': 'KEYWORD_SEARCH'}
        },
        {
            'name': 'MPN Exact Match',
            'params': {'mftrPartNumber': mpn}
        },
        {
            'name': 'Manufacturer Part Number',
            'params': {'manufacturerPartNumber': mpn}
        },
        {
            'name': 'Exact Match Field',
            'params': {'exactMatch': mpn}
        }
    ]
    
    results = []
    
    for method in search_methods:
        result = test_webapp_search(region, mpn, method['params'])
        print_result(method['name'], result)
        
        if result['success']:
            results.append((region['name'], method['name'], result))
    
    return results


def test_rest_api_all(mpn):
    """Test REST API with all search methods"""
    
    print(f"\n{'='*80}")
    print(f"🌐 REST API v3 - {mpn}")
    print(f"{'='*80}")
    
    search_methods = [
        ('Keyword', 'keyword'),
        ('MPN Exact (manuPartNum:)', 'mpn_exact'),
        ('MPN Field', 'mpn_field')
    ]
    
    results = []
    
    for name, method in search_methods:
        result = test_rest_api(mpn, method)
        print_result(name, result)
        
        if result['success']:
            results.append(('REST API', name, result))
    
    return results


def main():
    """Main test execution"""
    
    print(f"\n{'='*80}")
    print(f"Farnell API Region & Search Method Test")
    print(f"{'='*80}\n")
    
    # Test with common components
    test_mpns = [
        'LM317',
        'ATMEGA328P-PU',
        'LM317MBSTT3G',
        'STM32F407VGT6'
    ]
    
    all_successes = []
    
    for mpn in test_mpns:
        print(f"\n\n{'#'*80}")
        print(f"# Testing MPN: {mpn}")
        print(f"{'#'*80}")
        
        # Test webapp endpoints for each region
        for region in REGIONS:
            successes = test_region(region, mpn)
            all_successes.extend(successes)
        
        # Test REST API
        rest_successes = test_rest_api_all(mpn)
        all_successes.extend(rest_successes)
    
    # Print summary
    print(f"\n\n{'='*80}")
    print(f"📋 SUMMARY")
    print(f"{'='*80}\n")
    
    if all_successes:
        print(f"✅ Found {len(all_successes)} working combination(s):\n")
        
        for region, method, result in all_successes:
            print(f"   ✅ {region} - {method}: {result['num_results']} products")
        
        print(f"\n💡 Recommended configuration:")
        best = all_successes[0]
        print(f"   Region: {best[0]}")
        print(f"   Method: {best[1]}")
        print(f"   Results: {best[2]['num_results']}")
        
        if 'url' in best[2]:
            print(f"\n📝 Example URL:")
            print(f"   {best[2]['url']}")
    else:
        print(f"❌ No working combinations found!\n")
        print(f"⚠️ Possible issues:")
        print(f"   1. Invalid API key (check FARNELL_API_KEY)")
        print(f"   2. API key not activated (check Farnell developer portal)")
        print(f"   3. Wrong product database (try different regions)")
        print(f"   4. MPN format mismatch (check exact MPN formatting)")
        print(f"   5. Rate limiting (wait and retry)")
        
        print(f"\n💡 Next steps:")
        print(f"   1. Verify API key at: https://partner.element14.com/")
        print(f"   2. Check API key permissions (need 'Product Search' access)")
        print(f"   3. Try searching on website first to verify product exists")
        print(f"   4. Contact Farnell support if API key is correct")


if __name__ == '__main__':
    main()

File: scripts/test_farnell_regions.mjs
/**
 * Тест Farnell API с разными вариантами поиска
 */

import { farnellByMPN, farnellByKeyword } from './src/integrations/farnell/client.mjs';
import dotenv from 'dotenv';

dotenv.config();

const FARNELL_KEY = process.env.FARNELL_API_KEY;
const FARNELL_REGION = process.env.FARNELL_REGION || 'uk.farnell.com';

// Проверим разные регионы и популярные товары
const TEST_REGIONS = ['uk.farnell.com', 'us.newark.com'];
const TEST_PRODUCTS = [
  { name: 'Arduino Uno', mpn: 'A000066', keyword: 'Arduino Uno' },
  { name: 'Raspberry Pi 4', mpn: 'RPI4-MODBP-8GB', keyword: 'Raspberry Pi 4' },
  { name: 'STM32', mpn: 'STM32F407VGT6', keyword: 'STM32F407' },
  { name: 'ATMega328', mpn: 'ATMEGA328P-PU', keyword: 'ATMEGA328' },
];

async function testFarnellProduct(region, product, searchType = 'mpn') {
  const searchFunc = searchType === 'mpn' ? farnellByMPN : farnellByKeyword;
  const searchValue = searchType === 'mpn' ? product.mpn : product.keyword;
  
  console.log(`\n   ${searchType.toUpperCase()}: ${searchValue}`);
  
  try {
    const result = await searchFunc({
      apiKey: FARNELL_KEY,
      region: region,
      q: searchValue,
      limit: 5
    });
    
    if (!result.ok) {
      console.log(`   ❌ API Error: ${result.status}`);
      return { ok: false, status: result.status };
    }
    
    const productCount = result.data?.premierFarnellPartNumberReturn?.numberOfResults || 0;
    console.log(`   📦 Найдено: ${productCount}`);
    
    if (productCount === 0) {
      return { ok: false, count: 0 };
    }
    
    const products = result.data.premierFarnellPartNumberReturn?.products || [];
    if (products.length > 0) {
      const p = products[0];
      console.log(`   ✅ ${p.sku} - ${p.displayName}`);
      console.log(`      Brand: ${p.brandName}`);
      console.log(`      Attributes: ${p.attributes?.length || 0}`);
      console.log(`      Images: ${p.images?.length || 0}`);
      
      return { 
        ok: true, 
        count: productCount,
        specs: p.attributes?.length || 0,
        images: p.images?.length || 0
      };
    }
    
    return { ok: false, count: productCount, noProducts: true };
    
  } catch (err) {
    console.log(`   ❌ Exception: ${err.message}`);
    return { ok: false, error: err.message };
  }
}

async function main() {
  console.log('🟠 Тестирование Farnell API\n');
  console.log(`API Key: ${FARNELL_KEY?.substring(0, 10)}...`);
  console.log('='.repeat(60));
  
  const results = {};
  
  for (const region of TEST_REGIONS) {
    console.log(`\n🌍 Регион: ${region}`);
    results[region] = {};
    
    for (const product of TEST_PRODUCTS) {
      console.log(`\n📦 ${product.name}`);
      
      // Пробуем и MPN и keyword
      const mpnResult = await testFarnellProduct(region, product, 'mpn');
      const keywordResult = await testFarnellProduct(region, product, 'keyword');
      
      results[region][product.name] = {
        mpn: mpnResult,
        keyword: keywordResult
      };
    }
  }
  
  // Анализ
  console.log('\n' + '='.repeat(60));
  console.log('📊 ИТОГИ\n');
  
  for (const region of TEST_REGIONS) {
    const successMPN = Object.values(results[region])
      .filter(r => r.mpn.ok).length;
    const successKeyword = Object.values(results[region])
      .filter(r => r.keyword.ok).length;
      
    console.log(`\n🌍 ${region}:`);
    console.log(`   MPN search: ${successMPN}/${TEST_PRODUCTS.length} успешно`);
    console.log(`   Keyword search: ${successKeyword}/${TEST_PRODUCTS.length} успешно`);
    
    if (successMPN > 0 || successKeyword > 0) {
      console.log(`   ✅ Регион работает!`);
      
      // Покажем что нашли
      Object.entries(results[region]).forEach(([name, result]) => {
        if (result.mpn.ok) {
          console.log(`      ${name}: ${result.mpn.specs} specs, ${result.mpn.images} images (MPN)`);
        } else if (result.keyword.ok) {
          console.log(`      ${name}: ${result.keyword.specs} specs, ${result.keyword.images} images (Keyword)`);
        }
      });
    } else {
      console.log(`   ❌ Ничего не найдено`);
    }
  }
  
  console.log('\n' + '='.repeat(60));
  console.log('🎯 РЕКОМЕНДАЦИЯ:\n');
  
  const bestRegion = TEST_REGIONS.find(region => {
    const successMPN = Object.values(results[region]).filter(r => r.mpn.ok).length;
    const successKeyword = Object.values(results[region]).filter(r => r.keyword.ok).length;
    return successMPN > 0 || successKeyword > 0;
  });
  
  if (bestRegion) {
    console.log(`   Использовать регион: ${bestRegion}`);
    
    const mpnWorks = Object.values(results[bestRegion]).some(r => r.mpn.ok);
    const keywordWorks = Object.values(results[bestRegion]).some(r => r.keyword.ok);
    
    if (mpnWorks && keywordWorks) {
      console.log(`   Стратегия: попробовать MPN first, fallback на keyword`);
    } else if (mpnWorks) {
      console.log(`   Стратегия: только MPN search`);
    } else {
      console.log(`   Стратегия: только keyword search`);
    }
  } else {
    console.log(`   ❌ Farnell API не работает ни в одном регионе`);
    console.log(`   Проверить: API key, rate limits, доступность сервиса`);
  }
}

main().catch(console.error);
